V.2

BridgeAR opened the pull request:
Please have a look at the individual commits for a more detailed description.

BridgeAR commented:
I rewrote the whole logic and added a completely new algorithm for the stable part that outperforms any to me known alternative.

BridgeAR commented:
I am not sure how to enter the type information by the way - because it is a function including a property. So it is something in-between a interface and a declaration.

davidmarkclements commented on readme.md:
```diff
@@ -1,88 +1,122 @@
-usage)
+Gracefully handles circular structures instead of throwing.
+
+Provides a stable version as well that will also gracefully handle circular
```
"stable version" is ambiguous (immediately makes you think software versioning)

how about

Provides additional optional sorted key functionality (stable sorted JSON), which is also fast and safe

davidmarkclements commented on readme.md:
```diff
@@ -1,88 +1,122 @@
-`fast-safe-stringify` would not attempt to detect circular dependencies on
-objects that have a `toJSON` function. If you need to do that, you will need to
-attach a `toJSON.forceDecirc = true` property, like so:
+Using the stable version also works the same as [JSON.stringify][]:
```
see prior note on "stable version"

davidmarkclements commented on readme.md:
```diff
@@ -1,88 +1,122 @@
+
+[JSON.stringify][]'s `replacer` and `space` options are supported and work the
+same as JSON.stringify with one exception: in case a circular structure is
+detected the replacer will receive the string `[Circular]` as argument instead
```
typo: as **the** argument

davidmarkclements commented:
for curiosity/posterity, could you please also post comparison benchmarks (master vs pr branch)

BridgeAR commented:
Old
```
fastSafeStringifyBench*10000: 37.160ms
fastSafeStringifyCircBench*10000: 55.967ms
fastSafeStringifyDeepBench*10000: 414.544ms
fastSafeStringifyDeepCircBench*10000: 440.958ms
fastSafeStringifyBench*10000: 32.556ms
fastSafeStringifyCircBench*10000: 45.782ms
fastSafeStringifyDeepBench*10000: 428.917ms
fastSafeStringifyDeepCircBench*10000: 426.085ms
```
New
```
fastSafeStringifyBench*10000: 32.890ms
fastSafeStringifyCircBench*10000: 39.071ms
fastSafeStringifyDeepBench*10000: 357.374ms
fastSafeStringifyDeepCircBench*10000: 370.444ms
fastSafeStringifyBench*10000: 25.070ms
fastSafeStringifyCircBench*10000: 36.617ms
fastSafeStringifyDeepBench*10000: 352.785ms
fastSafeStringifyDeepCircBench*10000: 358.732ms
```

davidmarkclements commented:
👏

mcollina commented:
Nice! I don’t understand the change exactly, what is the stable version?

pino-noir should be checked, it was relying on internals of this library to do its job.

davidmarkclements commented:
ok @mcollina not knowing "stable version" definitely indicates this needs more explanation, 
and possibly a method rename

@mcollina see https://www.npmjs.com/package/json-stable-stringify

davidmarkclements commented:
really good point re pino-noir - @BridgeAR there's a type coupling between these two libraries, although from some output you showed me earlier it seems like you've already been working with pino-noir ?

BridgeAR commented:
I updated the documentation a bit.

davidmarkclements commented:
In terms of the d.ts file, I think this doc has details on how to implement for function-objects

https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-function-d-ts.html

davidmarkclements commented:
@BridgeAR – get job clarifying in docs update

can we rename the `stable` method to  `stableStringify` (or even aliasing is fine)

Two reasons: 

1. it's clearer
2. allows for:

```js
const { stableStringify } = require('fast-safe-stringify')
```

BridgeAR commented:
Hmm... this will indeed break pino-noir... :cry: I will have to investigate how to circumvent that...
The issue is that it is expected that pino-noir is able to manipulate the object that is passed to `JSON.stringify`. It will try to add properties to the `circular` object. And here it breaks because the way this code works is to replace the value with the string directly and replace it back to the original object before finishing...

Besides that: I am fine with renaming the function name.

BridgeAR commented:
I do not yet know how to solve this.

mcollina commented:
This is semver-major anyway, we would have to so a semver-major in pino and pino-noir too.
It might take a bit more time to roll this out in pino.

davidmarkclements commented:
the main thing we want to avoid is slowing pino-noir down

worst case scenario, we keep pino-noir on version 1 - it doesn't need the additional features

BridgeAR commented:
The stable version passes all tests as it circumvents the problem.
Doing something similar lets the pino tests pass but one stringify test fails and fixing that would be performance penalty again.

I am going to look into it. I might also have a look at pino if you like me to do that.

davidmarkclements commented:
absolutely!

BridgeAR commented:
I added a few more commits and highlighted the differences to `JSON.stringify` better.

I also added a alias and I made the tests a bit nicer.

When looking in how I can fix the breaking behavior for pino-noir I found that it would be possible to use a object with `toJSON` that returns the necessary string. That makes almost all tests pass (since it is a object again, it is possible to add properties on it again) in pino-noir but not all. I tried to play around with that a bit further but since it would not fix everything I decided against changing that.

It is just not possible to manipulate the input value while using `toJSON` or the `replacer`.

mcollina commented:
Can you add a `.travis.yml`?

BridgeAR commented:
With the new benchmarks it is clear that this will only improve the performance for simple objects. If the objects gets more complicated it will not work anymore.

New
```
fast-safe-stringify:   simple object x 190,757 ops/sec ±1.93% (89 runs sampled)
fast-safe-stringify:   circular      x 103,776 ops/sec ±7.74% (81 runs sampled)
fast-safe-stringify:   deep          x 25,483 ops/sec ±2.61% (92 runs sampled)
fast-safe-stringify:   deep circular x 26,030 ops/sec ±0.89% (92 runs sampled)
```
Against Old
```
fast-safe-stringify:   simple object x 117,478 ops/sec ±1.61% (89 runs sampled)
fast-safe-stringify:   circular      x 70,008 ops/sec ±0.93% (95 runs sampled)
fast-safe-stringify:   deep          x 33,074 ops/sec ±0.67% (92 runs sampled)
fast-safe-stringify:   deep circular x 30,894 ops/sec ±1.13% (91 runs sampled)
```
Other old
```
fast-safe-stringify:   simple object x 673,063 ops/sec ±1.76% (88 runs sampled)
fast-safe-stringify:   circular      x 349,338 ops/sec ±1.26% (94 runs sampled)
fast-safe-stringify:   deep          x 40,231 ops/sec ±1.74% (90 runs sampled)
fast-safe-stringify:   deep circular x 36,310 ops/sec ±2.10% (89 runs sampled)
```
Against other new
```
fast-safe-stringify:   simple object x 984,145 ops/sec ±1.07% (93 runs sampled)
fast-safe-stringify:   circular      x 502,014 ops/sec ±0.94% (95 runs sampled)
fast-safe-stringify:   deep          x 29,924 ops/sec ±0.56% (93 runs sampled)
fast-safe-stringify:   deep circular x 29,272 ops/sec ±1.06% (92 runs sampled)
```

BridgeAR commented:
Ignore my last comment. Seems like a background task was running and ruined the results... you could of course also run the benchmarks again on your side.

BridgeAR commented:
This should be ready. @mcollina @davidmarkclements would you be so kind and have another look?

BridgeAR commented:
Rebased

davidmarkclements commented:
great work @BridgeAR – published v2.0.0

now let's see what we can do with pino-noir
